### PR TITLE
Windows

### PR DESCRIPTION
**Pull request by pukkamustard**
Adds support for cross-compiling from Linux to Windows and releasing for Windows from Linux.

Tested on virtual machine with Play but without real Senso. Further in-field testing should be done.

## Todo

- [x] Test RFID functionality on Windows @knuton 
- [x] Test Authenticode signature
- [x] Run as Windows Service
- [x] Test with real hardware

### CODE REVIEW

**Comment by pukkamustard** [comment]
This PR also makes travis run the `crossbuild` make target (which tries to build for all supported platforms). This is a good thing but increases time Travis requires considerably as it needs to recompile the cross-compilation toolchains on every invocation.

**Comment by knuton** [comment]
Did you test the Authenticode signature (codesigning) on the Windows system?

**Comment by pukkamustard** [comment]
> Did you test the Authenticode signature (codesigning) on the Windows system?

I can run the executable. But don't know what the expected behaviour is (i.e. what popups should or should not show up).

**Comment by pukkamustard** [comment]
> This PR also makes travis run the crossbuild make target (which tries to build for all supported platforms). This is a good thing but increases time Travis requires considerably as it needs to recompile the cross-compilation toolchains on every invocation.

Using Travis cache enabled (see https://github.com/travis-ci/travis-ci/issues/6604 and https://docs.travis-ci.com/user/caching/). However does not seem to be working for pull requests. I think it needs to be merged back into `develop` of the `dividat` repo and then travis will start pushing caches after builds. I hope it will start working when this gets merged.

**Comment by pukkamustard** [comment]
Windows service works like a charm. Added a note to readme on how to set it up.

For real deployments this will be automated using [Boxstarter](https://boxstarter.org/WhyBoxstarter) et al.

**Comment by knuton** [comment]
I haved tested RFID on Windows, it works just fine.

I also checked the Authenticode signature with `Get-AuthenticodeSignature` in PowerShell, which says it is valid.

However, the program was not treated kindly when it had the "Mark-of-the-Web".

Steps:

1. Ran unsigned binary (`make crossbuild`) transported via USB stick. :heavy_check_mark: 
2. Ran signed binary (`make release`) transported via USB stick. :heavy_check_mark: 
3. Ran signed binary with MotW downloaded in Chrome. :x: 
  - Chrome warned that it is a "rare download" that might be harmful.
  - Windows Defender quarantined the file I downloaded, flagged it as `Trojan:Win32/Sprisky.V!cl` which it detects using "AI". This category of malware is entirely for dynamically classified threats.
  - Windows Defender then also quarantined the same binary from the USB stick (even though it had no mark of the web).

I can not find information about the criteria used for classification.

**Comment by knuton** [comment]
Can you repeat your check but with the program having the Mark of the Web?

**Comment by pukkamustard** [comment]
Repeated test with "Mark of the Web" (download via Chromium). Two possibilities:
- If run the application from a PowerShell: runs without warning. :heavy_check_mark: 
- If double-click on application (which starts it in a Cmd.exe): Popup with warning (but nothing blocking a la Defender).

I am not sure I was able to reproduce the behavior you describe (Defender quarantining). It might be due to my system not running Defender (The only thing I find when searching for "Defender" is "Defender Firewall").

In any case. I have added a PowerShell script that should circumvent the problem. The script downloads latest version of application and installs it as service.

To test the script:

- `make release`
- Run a small webserver in repo root.
- Modify `install.ps1` to something like this:
  ```
  $releaseUrl = "http://192.168.1.21:9999/release/"
  $channel = "windows"
  ```
- Run the command in a PowerShell (Administrator):
  ```
  Set-ExecutionPolicy Bypass -Scope Process -Force; iex ((New-Object System.Net.WebClient).DownloadString('http://192.168.1.21:9999/install.ps1'))
   ```

@knuton Can you check if this circumvents the Mark of the Web problem?

**Comment by pukkamustard** [comment]
Also just checked remote logging: works fine.

I noticed this log entry:
```
    "msg": "Could not create smart card context.",
    "_application": "driver",
    "package": "rfid",
    "os": "windows",
    "level": "error",
    "error": "scard: error(8010001d): The Smart Card Resource Manager is not running.",
    "version": "2.1.0-rc1-23-g7555f16",
    "releaseChannel": "windows",
```

Expected? Is there something else that needs to be running for it to work?

File: Makefile [hunk 1]
```diff
@@ -93,22 +97,22 @@ $(LINUX_RELEASE): $(LINUX_BIN)
...
-		#-h sha1 \
-		#-n "Dividat Driver" \
-		#-i "https://www.dividat.com/" \
-		#-in $(WINDOWS_RELEASE) \
-		#-out $(WINDOWS_RELEASE)
+WINDOWS_RELEASE = $(RELEASE_DIR)/dividat-driver-windows-amd64-$(VERSION).exe
+$(WINDOWS_RELEASE): $(WINDOWS_BIN)
+	cp $(WINDOWS_BIN) $(WINDOWS_RELEASE)
+	upx $(WINDOWS_BIN)
+	osslsigncode sign \
```
Comment (knuton) [inline Makefile]: Timestamp missing. driverv0 should give details to use.

Further information:
- https://docs.microsoft.com/en-us/windows/desktop/seccrypto/time-stamping-authenticode-signatures#about-cryptographic-time-stamping
- https://github.com/antoinevg/osslsigncode/blob/master/README#L73-L76

File: install.ps1 [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# This script will download and install Dividat Driver as a Windows service.
+
+## Configuration ##########################################
+$releaseUrl = "https://dist.dividat.com/releases/driver2/"
+$channel = "master"
+$installDir = "C:\dividat-driver"
```
Comment (knuton) [inline install.ps1]: Default should be "C:\Program Files\dividat-driver".

**Comment by pukkamustard** [comment]
Requested changes made.

Tested on a real Windows 10 with a real Senso. Tip top... going ahead and merging this.

**Comment by pukkamustard** [comment]
[Driver: Windows](https://trello.com/c/0cNpkJ36/168-driver-windows)

**Pull request merged**